### ISSUE TITLE
Rc/Arc equality checking could short-circuit on equal pointers if T: Eq

### ISSUE THREAD

**Issue by joliss** [rust-lang/rust#42655]
The `PartialEq` implementations for `Rc` and `Arc` currently always defer to the equality implementations for their inner values ([rc.rs:770](https://github.com/rust-lang/rust/blob/554c685b0b3b25b7aa752717edf50b8d6bcab7a0/src/liballoc/rc.rs#L770-L788)):

```rust
    fn eq(&self, other: &Rc<T>) -> bool {
        **self == **other
    }
```

However, where `T: Eq`, we may assume that the inner values obey reflexivity (`a == a`), and so we could short-circuit the comparison if the pointers are equal. That is, instead of checking `**self == **other`, we could check `self.ptr_eq(other) || **self == **other`.

It is conceivable that in rare cases the equality-check on the inner value is so cheap that attempting to short-circuit adds more overhead than it saves. However, my sense is that `Rc` is usually used on types that are too complex to clone cheaply, and so it stands to reason that in most cases, the equality check on the inner value is expensive. (This is certainly true for the use cases I've encountered.) Checking equality on the pointers furthermore is very cheap, since we're dereferencing them anyway.

I'm cautiously volunteering a pull request if people think that this is a reasonable change. *Update: Pull request at #42965.*

**Comment by joliss** [comment]
Hm, I believe this is actually impossible to implement until we have support for specialization (https://github.com/rust-lang/rust/issues/31844). Please correct me if I'm wrong.

**Comment by sfackler** [comment]
Specialization already exists - it's just unstable. That doesn't matter for use in the standard library, though.

**Comment by joliss** [comment]
Cool, thanks @sfackler. I posted a pull request at #42965.

**Comment by dtolnay** [comment]
Tagging C-feature-accepted based on the discussion in #42965. Someone will need to rebase the PR and address the review comments.

**Comment by mbrubeck** [comment]
This optimization was implemented by servo/servo#18793 in a fork of `Arc` used in Servo and Firefox.

It occurs to me that if this is valid for `Rc<T>` and `Arc<T>` then it should also be valid for `&T`.

**Comment by jdm** [comment]
The &T case was discussed in #42965.
